graphix/open_graph.py

```diff
@@ -0,0 +1,379 @@
+                continue
+
+            # g.phase() may be a fractions.Fraction object, but Measurement
+            # expects a float
+            measurements[v] = Measurement(float(g.phase(v)), "XY")
+
+        return cls(g_nx, measurements, inputs, outputs)
+
+    @property
+    def inputs(self) -> list[int]:
+        """Returns the inputs of the graph sorted by their position.
+
+        Example
+        ------
+        >>> import networkx as nx
+        >>> from graphix.open_graph import OpenGraph, Measurement
+        >>>
+        >>> g = nx.Graph([(0, 1), (1, 2), (2, 0)])
+        >>> measurements = {i: Measurement(0.5 * i, "XY") for i in range(2)}
+        >>> inputs = [0]
+        >>> outputs = [2]
+        >>>
+        >>> og = OpenGraph(g, measurements, inputs, outputs)
+        >>> og.inputs == inputs
+        True
+        """
+        inputs = [i for i in self.inside.nodes(data=True) if "is_input" in i[1]]
+        sorted_inputs = sorted(inputs, key=lambda x: x[1]["input_order"])
+        input_node_ids = [i[0] for i in sorted_inputs]
+        return input_node_ids
+
+    @property
+    def outputs(self) -> list[int]:
+        """Returns the outputs of the graph sorted by their position.
+
+        Example
+        ------
+        >>> import networkx as nx
+        >>> from graphix.open_graph import OpenGraph, Measurement
+        >>>
+        >>> g = nx.Graph([(0, 1), (1, 2), (2, 0)])
+        >>> measurements = {i: Measurement(0.5 * i, "XY") for i in range(2)}
+        >>> inputs = [0]
+        >>> outputs = [2]
+        >>>
+        >>> og = OpenGraph(g, measurements, inputs, outputs)
+        >>> og.outputs == outputs
+        True
+        """
+        outputs = [i for i in self.inside.nodes(data=True) if "is_output" in i[1]]
+        sorted_outputs = sorted(outputs, key=lambda x: x[1]["output_order"])
+        output_node_ids = [i[0] for i in sorted_outputs]
+        return output_node_ids
+
+    @property
+    def measurements(self) -> dict[int, Measurement]:
+        """Returns a dictionary which maps each node to its measurement. Output
+        nodes are not measured and hence are not included in the dictionary.
+
+        Example
+        ------
+        >>> import networkx as nx
+        >>> from graphix.open_graph import OpenGraph, Measurement
+        >>>
+        >>> g = nx.Graph([(0, 1), (1, 2), (2, 0)])
+        >>> measurements = {i: Measurement(0.5 * i, "XY") for i in range(2)}
+        >>> inputs = [0]
+        >>> outputs = [2]
+        >>>
+        >>> og = OpenGraph(g, measurements, inputs, outputs)
+        >>> og.measurements == {
+        ...     0: Measurement(0.0, "XY"),
+        ...     1: Measurement(0.5, "XY"),
+        ... }
+        True
+        """
+        return {n[0]: n[1]["measurement"] for n in self.inside.nodes(data=True) if "measurement" in n[1]}
+
+    def perform_z_deletions_in_place(self):
+        """Removes the Z-deleted nodes from the graph in place"""
+        z_measured_nodes = [
```

Comment:
Please use genexp if you're using it only for iteration.

Reply:
I can't do that here because I am removing nodes later on. If I use a genexp then it will complain that the size of `.inside.nodes` is changing during iteration.